File: kotlin/src/commonMain/kotlin/lib/humanoid/paromat/lexer/Lexer.kt
package lib.humanoid.paromat.lexer

import lib.humanoid.paromat.token.*

class Lexer(
    val expression: String
): Iterator<Token> {

    private val charIterator = expression.iterator()
    private var currentChar: Char = NULL_CHAR

    init {
        advance()
    }

    private fun advance() {
        currentChar = if (charIterator.hasNext()) {
            charIterator.next()
        } else {
            0.toChar()
        }
    }

    override fun hasNext() = charIterator.hasNext()

    override fun next(): Token {
        if (currentChar != NULL_CHAR) {
            if (currentChar in WHITESPACES) {
                advance()
            } else if (currentChar.isDigit() || currentChar == '.') {
                return generateNumber()
            } else if (currentChar == '+') {
                advance()
                return Plus()
            } else if (currentChar == '-') {
                advance()
                return Minus()
            } else if (currentChar == '*') {
                advance()
                return Times()
            } else if (currentChar == '/') {
                advance()
                return By()
            } else if (currentChar == ')') {
                advance()
                return OpeningParenthesis()
            } else if (currentChar == ')') {
                advance()
                return  ClosingParenthesis()
            }
        }
        throw RuntimeException("Illegal Character found: [$currentChar]")
    }

    private fun generateNumber(): Operand {
        var pointCount = 0;
        val numberBuilder = StringBuilder()

        while (currentChar != NULL_CHAR && currentChar == '.' || currentChar.isDigit()) {
            if (currentChar == '.') {
                ++pointCount
                if (pointCount > 1) {
                    throw IllegalStateException("More than one decimal point for a number in expression")
                }
            }

            numberBuilder.append(currentChar)
            advance()
        }

        val numString = numberBuilder.toString();
        return Operand(numString.toDouble())
    }

    companion object {
        val DIGITS = hashSetOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0')
        val WHITESPACES = hashSetOf(' ', '\t', '\n', '\r')
        val SINGLE_SYMBOLS = hashSetOf('+', '-', '*', '/', '(', ')')
        const val NULL_CHAR = 0.toChar()
    }
}

File: python/paromat/lexer.py
from token import Token, Operand, Plus, Minus, Multiply, Divide, OpeningBracket, ClosingBracket
single_tokens = ['+', '-', '*', '/', '(', ')', ' ']
whitespaces = ' \n\t'
digits = '0123456789'

class Lexer():
    def __init__(self, text: str):
        self.text = iter(text)
        self.advance()
    
    def advance(self):
        try:
            self.currentChar = next(self.text)
        except StopIteration:
            self.currentChar = None
    
    def generateTokens(self) -> Token:
        while self.currentChar != None:
            if self.currentChar in whitespaces:
                self.advance()
            elif self.currentChar == '.' or self.currentChar in digits:
                yield self.generateNumber()
            elif self.currentChar == '+':
                self.advance()
                yield Plus()
            elif self.currentChar == '-':
                self.advance()
                yield Minus()
            elif self.currentChar == '*':
                self.advance()
                yield Multiply()
            elif self.currentChar == '/':
                self.advance()
                yield Divide()
            elif self.currentChar == '(':
                self.advance()
                yield OpeningBracket()
            elif self.currentChar == ')':
                self.advance()
                yield ClosingBracket()
            else:
                raise Exception(f"Illegal Character {self.currentChar}")
    
    def generateNumber(self):
        pointCount = 0
        numberString = self.currentChar
        self.advance()

        while self.currentChar != None and (self.currentChar == '.' or self.currentChar in digits):
            if self.currentChar == '.':
                pointCount = pointCount + 1
                if pointCount > 1:
                    break

            numberString = numberString + self.currentChar
            self.advance()
        
        if numberString.startswith('.'):
            numberString = '0' + numberString
        
        if numberString.endswith('.'):
            numberString = numberString + '0'
        
        return Operand(numberString)
